pkg/pillar/utils/generics.go

```diff
@@ -138,3 +140,19 @@ func FilterDuplicatesFn[Type any](list []Type, equal func(a, b Type) bool) (filt
 	}
 	return
 }
+
+// ExtractFields extracts all fields from the given type and its anonymous fields
+// and adds them to the given map. This can be used for a careful deserialization.
+func ExtractFields(t reflect.Type, fieldMap *map[string]bool) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+
+		if field.Anonymous {
+			// If this is an anonymous field, recursively extract its fields as well.
+			ExtractFields(field.Type, fieldMap)
```

Comment:
I realize you don't need this nesting now, but when we do I think the field names can conflict.

If you have e.g.,
type top struct {
    a int
    b nest
}
type next struct {
   a int
}

Then field.Name will be "a" for both top.a and top.nest.a
One could fix this by adding a prefix argument to ExtractFields so be able to e.g., construct these dotten names I used as an example.
Or just handle the top level fields here??

Some other alternatives:
For the critical fields one can do something simpler by checkig that their values are not default aka empty. But that doesn't allow warning for extraneous fields. One can call json.Decode a second time setting the flag (I forget exactly where that needs to be set) to error on unknown fields, but I don't know if that will return an error string indocating the names of the extraneous fields, or just the first extra field.

Reply:
Just in case, here is a link for a snippet that helps understand how it works:
https://go.dev/play/p/g7n6mn8nFnA

Reply:
Maybe I misunderstood anynomous to mean 'unexported'. If it is instead what we have with a type without a name, then what you have is perfectly fine.

Reply:
Yep, a field with a type but without a name is an anonymous field (or an embedded field) in terms of Go =)